src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1408,46 +1413,105 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
+                        throw new Exception("unkown ast type");
                     }
-                    if (argRank != GetRank(p)) return false;
+                    if (!DoesParamRankMatchArg(fep, p, t, argIndex)) return false;
                 }
-                else if (args[argIndex] is IdentifierNode idn)
+                else if (currentArg is IdentifierNode idn)
                 {
                     if (idn.ReplicationGuides.Count > 0) return false;
 
                     var t = variables[idn.Value].Item2;
                     if (t == typeof(object)) return false;
 
-                    if (!p.Equals(t))
+                    if(!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
+                }
+                else if (currentArg is RangeExprNode rgnNode)
+                {
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(rgnNode.ID, out t))
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
-        private static int GetRank(Type type)
+        /// <summary>
+        /// Check that a parameter's rank and type are compatible with the given argument.
+        /// This function besides checking rank directly, also does checks using the FEP attributes.
+        /// </summary>
+        /// <param name="fep">the function end point wrapper</param>
+        /// <param name="paramType">type of param on the fep</param>
+        /// <param name="argType">actual argument type</param>
+        /// <param name="paramIndex">parameter index in list of params to fep.</param>
+        /// <returns></returns>
+        private static bool DoesParamRankMatchArg(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)
+        {
+            if (!paramType.Equals(argType))
+            {
+                //if both types are value types, let's
+                //let coercion figure this out.
+                if (paramType.IsValueType && argType.IsValueType)
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(fep,argType,paramIndex);
+                var paramRank = GetRank(fep,paramType, paramIndex);
+
+                //if the param is an arbitrary rank array and
+                //arg is some array type, replication should not occur.
+                if (paramRank == -1 && argRank > 0)
```

Comment:
yes, rank 0 would just be a single item I believe.

One thing that is unclear to me -

If a parameter of rank -1 (arbitrary) will never replicate with a single item OR an array, then this check is unnecessary.

@aparajit-pratap is that always the case?

Reply:
Comparing a single parameter to an argument, if the parameter has the arbitrary rank attribute, then it doesn't matter what rank the argument has, it will not replicate. However, for a function with more than one parameter, there can be one parameter with the attribute, and another one that doesn't. In this case, the function can still replicate over the parameter that doesn't have the attribute. So only if `DoesParamRankMatchArg` returns true for all parameter-argument pairs for a function call, then we can decide not to replicate.

Reply:
ahh, good call, let me see if I can write a test for that case and address it.